Update MurmurHash3_x86_32 to use the cuco equivalent implementation

PointKernel opened the pull request:
## Description
This PR modifies `MurmurHash3_x86_32` to utilize the cuco equivalent implementation, eliminating duplication. Additionally, it resolves a minor issue in `xxhash_64` where the seed was not correctly passed to the underlying implementation.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

bdice commented on cpp/include/cudf/hashing/detail/default_hash.cuh:
```diff
@@ -16,7 +16,7 @@
 #pragma once
 
-#include <cudf/hashing/detail/murmurhash3_x86_32.cuh>
+#include "murmurhash3_x86_32.cuh"
```
From our developer guide:

> Use `<>` for all includes except for internal headers that are not in the include directory.

https://github.com/rapidsai/cudf/blob/branch-25.02/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#includes

bdice replied:
```suggestion
#include <cudf/hashing/detail/murmurhash3_x86_32.cuh>
```

bdice commented on cpp/include/cudf/hashing/detail/murmurhash3_x86_32.cuh:
```diff
@@ -16,165 +16,91 @@
 
 #pragma once
 
+#include "hash_functions.cuh"
```
Revert to `<cudf/hashing/detail/hash_functions.cuh>`, see above.

bdice commented on cpp/include/cudf/hashing/detail/murmurhash3_x86_32.cuh:
```diff
@@ -16,165 +16,91 @@
-        h ^= k1;
-    };
-    return h;
+    return cuco::murmurhash3_32<Key>::operator()(key);
```
Is there a better way to write this so that we "just call the call operator" instead of using `::operator()`? Does this work?
```suggestion
    return cuco::murmurhash3_32<Key>{}(key);
```

bdice replied:
Ah. Is this a case where composition would be better than inheritance? https://en.wikipedia.org/wiki/Composition_over_inheritance

PointKernel replied:
> Is this a case where composition would be better than inheritance?

Great question! I'd love to hear @lamarrr's insights on this.

lamarrr replied:
The current approach looks okay to me, although I'd prefer having it as a member of the hasher if we were going for composition. It's purely stylistic in this case and gives no semantic difference.

bdice commented on cpp/include/cudf/hashing/detail/murmurhash3_x86_32.cuh:
```diff
@@ -16,165 +16,91 @@
   numeric::decimal32 const& key) const
 {
-  return compute(key.value());
+  auto const val = key.value();
+  auto const len = sizeof(val);
+  return this->compute_hash(reinterpret_cast<cuda::std::byte const*>(&val), len);
```
Can we just use the `operator()` of this class and pass `key.value()`? The previous code was simpler here.

PointKernel replied:
https://github.com/rapidsai/cudf/blob/07845438855c3787dc5daec0195bd75a399cd598/cpp/include/cudf/hashing/detail/murmurhash3_x86_32.cuh#L71-L75

It's actually doing the same as in the current implementation (see above). The issue is that `numeric::decimal32` and its underlying type are not the same, so the current implementation uses the above `compute` member function to handle key types that differ from `Key`. cuco achieves the same goal without relying on the error-prone `compute` member, preventing users from unintentionally passing arbitrary types to the hasher.

bdice replied:
Aaagh. I see why this is the case. But I wish we didn't need so much repetition of the `reinterpret_cast` pattern.

bdice reviewed (approved):
I guess this is fine, though I had to rethink my own views on how the hashers work. Thanks @PointKernel for explaining the pitfalls in my expectations.

lamarrr commented on cpp/include/cudf/hashing/detail/murmurhash3_x86_32.cuh:
```diff
@@ -24,157 +24,82 @@
 
-  template <typename T>
-  result_type __device__ inline compute(T const& key) const
+  template <typename Extent>
```
We should use `std::uint64_t` here and remove the template instantiation. It's unnecessary as all byte ranges will be within that range.

PointKernel commented:
@bdice @lamarrr Thank you both for the review! I’ve addressed all your comments in the PR and will open a follow-up PR to apply the same improvement to xxhash, will request your review there as well.

PointKernel commented:
/merge
